cpp/src/copying/pack.cpp

```diff
@@ -0,0 +1,143 @@
+  for (auto&& col : cols) {
+    add_column_to_vector(serialize(col, base_ptr), table_metadata);
+    std::vector<column_view> children;
+    for (size_t i = 0; i < col.num_children(); i++) { children.push_back(col.child(i)); }
+
+    serialize_columns(children, base_ptr, table_metadata);
+  }
+}
+
+}  // namespace
+
+packed_table pack(table_view const& input,
+                  cudaStream_t stream,
+                  rmm::mr::device_memory_resource* mr) {
+  contiguous_split_result contiguous_data =
+    std::move(detail::contiguous_split(input, {}, mr, stream).front());
+
+  serialized_column table_element = {{}, 0, 0, 0, contiguous_data.table.num_columns()};
+
+  auto result =
+    packed_table(std::make_unique<std::vector<uint8_t>>(), std::move(contiguous_data.all_data));
+  add_column_to_vector(table_element, result.table_metadata.get());
+
+  std::vector<column_view> table_columns(contiguous_data.table.begin(),
+                                         contiguous_data.table.end());
+
+  serialize_columns(table_columns,
+                    static_cast<uint8_t const*>(result.table_data->data()),
+                    result.table_metadata.get());
+
+  return result;
+}
+
+namespace {
+
+column_view deserialize_column(serialized_column serial_column,
+                               std::vector<column_view> const& children,
+                               uint8_t const* base_ptr) {
+  auto data_ptr = serial_column._data_offset != -1 ? base_ptr + serial_column._data_offset : 0;
+
+  // size_t is an unsigned int so -1 is the max value of size_t. If the offset
+  // is UINT64_MAX then just assume there's no null mask instead of thinking
+  // what if there IS a null mask but the buffer is just -1u sized. This translates
+  // to 16 EB of memory. No GPU has that amount of memory and it'll be a while
+  // before anyone does. By that time, we'll have bigger problems because all code
+  // that exists will need to be re-written to consider memory > 16 EB. It'll be
+  // bigger than Y2K; and I'll be prepared with a cottage in Waknaghat and a lifetime
+  // supply of soylent and shotgun ammo.
+  // TODO: Replace above with better reasoning
+  auto null_mask_ptr =
+    serial_column._null_mask_offset != -1
+      ? reinterpret_cast<bitmask_type const*>(base_ptr + serial_column._null_mask_offset)
+      : 0;
+
+  return column_view(serial_column._type,
+                     serial_column._size,
+                     data_ptr,
+                     null_mask_ptr,
+                     UNKNOWN_NULL_COUNT,
+                     0,
+                     children);
+}
+
+}  // namespace
+
+contiguous_split_result unpack(std::unique_ptr<packed_table> input) {
+  auto serialized_columns =
+    reinterpret_cast<serialized_column const*>(input->table_metadata->data());
+  uint8_t const* base_ptr = static_cast<uint8_t const*>(input->table_data->data());
+  size_t current_index    = 1;
+
+  std::function<std::vector<column_view>(size_type)> get_columns;
```

Comment:
Why a `std::function` instead of a lambda?

Reply:
Because `auto` type deduction doesn't work when the function is called recursively. https://stackoverflow.com/questions/2067988/recursive-lambda-functions-in-c11